File: samples/Foundgine.SupplyChain.Advanced/Semantic/Semantics/SupplyChainSemanticModel.cs
using Foundgine.Generated;
using Foundgine.Core.Abstractions;
using Foundgine.Core.Semantic.Metadata;
using Foundgine.Core.Semantic;

namespace Foundgine.SupplyChain.Advanced.Semantics;

/// <summary>
/// Application semantic configuration for the Supply Chain showcase.
/// Structural entities, fields, identities and direct relationships come from
/// generated Foundgine metadata. A small strongly typed manual semantic overlay
/// then enriches Product and ProductComponent, while this layer adds the
/// application-level logical traversals that metadata cannot infer.
/// </summary>
public static class SupplyChainSemanticModel
{
    public static IMetadataCatalog Metadata { get; } = GeneratedMetadata.Build();
    public static SemanticModel Model { get; } = Build();

    public static EntityId Product => Entity("Product");
    public static EntityId Component => Entity("ProductComponent");
    public static EntityId Supplier => Entity("Supplier");
    public static EntityId Shipment => Entity("Shipment");
    public static EntityId InventoryLot => Entity("InventoryLot");
    public static EntityId Warehouse => Entity("Warehouse");
    public static EntityId BusinessUnit => Entity("BusinessUnit");
    public static EntityId CustomerOrder => Entity("CustomerOrder");
    public static EntityId CustomerOrderLine => Entity("CustomerOrderLine");
    public static EntityId PurchaseOrder => Entity("PurchaseOrder");
    public static EntityId PurchaseOrderLine => Entity("PurchaseOrderLine");
    public static EntityId Certification => Entity("SupplierCertification");
    public static EntityId ComplianceIncident => Entity("ComplianceIncident");

    public static SemanticModel Build() =>
        Metadata.FromMetadata()
            .Overlay(ManualSupplyChainSemanticModel.Model)
            .Traversal("Product", "shipments", "purchaseOrderLines", "purchaseOrder", "shipments")
            .Traversal("Product", "supplierIncidents", "purchaseOrderLines", "purchaseOrder", "supplier", "incidents")
            .Build();

    public static FieldId Field(string entityName, string fieldName) =>
        Model.Get(Entity(entityName)).Fields.Single(field =>
            string.Equals(field.Name, fieldName, StringComparison.OrdinalIgnoreCase)).Id;

    public static RelationshipId Relationship(string entityName, string relationshipName) =>
        Model.Get(Entity(entityName)).Relationships.Single(relationship =>
            string.Equals(relationship.Name, relationshipName, StringComparison.OrdinalIgnoreCase)).Id;

    private static EntityId Entity(string entityName) =>
        Metadata.Entities.Single(entity =>
            string.Equals(entity.Name, entityName, StringComparison.OrdinalIgnoreCase)).EntityId;
}

File: src/java/samples/foundgine-supply-chain-advanced/src/main/java/com/foundgine/samples/supplychain/advanced/semantics/SupplyChainSemanticModel.java
package com.foundgine.samples.supplychain.advanced.semantics;

import com.foundgine.core.abstractions.*;
import com.foundgine.core.semantic.*;
import com.foundgine.core.semantic.metadata.*;
import com.foundgine.samples.supplychain.advanced.generated.GeneratedFoundgineMetadata;

/** Application semantic model for the Advanced Supply Chain sample. */
public final class SupplyChainSemanticModel {
    public static final IMetadataCatalog Metadata = GeneratedFoundgineMetadata.build();

    public static final EntityId PRODUCT = entity("Product");
    public static final EntityId COMPONENT = entity("ProductComponent");
    public static final EntityId SUPPLIER = entity("Supplier");
    public static final EntityId SHIPMENT = entity("Shipment");
    public static final EntityId INVENTORY_LOT = entity("InventoryLot");
    public static final EntityId WAREHOUSE = entity("Warehouse");
    public static final EntityId BUSINESS_UNIT = entity("BusinessUnit");
    public static final EntityId PURCHASE_ORDER = entity("PurchaseOrder");
    public static final EntityId PURCHASE_ORDER_LINE = entity("PurchaseOrderLine");
    public static final EntityId CERTIFICATION = entity("SupplierCertification");
    public static final EntityId COMPLIANCE_INCIDENT = entity("ComplianceIncident");

    public static final SemanticModel MODEL = build();

    public static SemanticModel build() {
        return SemanticModelDiscovery.fromMetadata(Metadata)
                .overlay(ManualSupplyChainSemanticModel.MODEL)
                .traversal(
                        "Product", "shipments", "purchaseOrderLines", "purchaseOrder", "shipments")
                .traversal(
                        "Product",
                        "supplierIncidents",
                        "purchaseOrderLines",
                        "purchaseOrder",
                        "supplier",
                        "incidents")
                .build()
                .freeze();
    }

    public static EntityId entity(String name) {
        for (var e : Metadata.entities()) {
            if (e.name().equalsIgnoreCase(name)) return e.entityId();
        }
        throw new IllegalStateException("Unknown metadata entity '" + name + "'.");
    }

    public static FieldId field(String entity, String field) {
        return MODEL.get(entity(entity)).fields().stream()
                .filter(f -> f.name().equalsIgnoreCase(field))
                .findFirst()
                .orElseThrow()
                .id();
    }

    public static RelationshipId relationship(String entity, String relationship) {
        return MODEL.get(entity(entity)).relationships().stream()
                .filter(r -> r.name().equalsIgnoreCase(relationship))
                .findFirst()
                .orElseThrow()
                .id();
    }

    private SupplyChainSemanticModel() {}
}
